File: src/diagnostics/report.py
"""
Formatador de relatorios de diagnostico: transforma a lista de issues
em texto legivel, tabelas resumidas, e sugestoes automaticas de tratamento.
"""


class DiagnosticReport:
    """Formata o resultado do scanner em relatorio legivel + sugestoes."""

    def __init__(self, issues: list):
        self.issues = issues

    def resumo(self) -> str:
        """Resumo geral: contagem por gravidade e categoria."""
        if not self.issues:
            return "  Nenhum problema encontrado. Base parece saudavel."

        criticos = [i for i in self.issues if i.gravidade == "critico"]
        avisos = [i for i in self.issues if i.gravidade == "aviso"]
        infos = [i for i in self.issues if i.gravidade == "info"]

        linhas = []
        linhas.append(f"  Total de problemas: {len(self.issues)}")
        linhas.append(f"    Criticos: {len(criticos)}")
        linhas.append(f"    Avisos:   {len(avisos)}")
        linhas.append(f"    Info:     {len(infos)}")

        categorias = {}
        for i in self.issues:
            categorias[i.categoria] = categorias.get(i.categoria, 0) + 1
        linhas.append("")
        linhas.append("  Por categoria:")
        for cat, count in sorted(categorias.items(), key=lambda x: -x[1]):
            linhas.append(f"    {cat}: {count}")

        return "\n".join(linhas)

    def detalhado(self) -> str:
        """Relatorio detalhado com todos os problemas."""
        if not self.issues:
            return "  Nenhum problema encontrado."

        linhas = []
        ordem_grav = {"critico": 0, "aviso": 1, "info": 2}
        issues_ordenados = sorted(self.issues, key=lambda x: ordem_grav.get(x.gravidade, 3))

        for i, issue in enumerate(issues_ordenados, 1):
            icone = {"critico": "[!!!]", "aviso": "[ ! ]", "info": "[ i ]"}
            linhas.append(f"  {i:>3}. {icone.get(issue.gravidade, ' ? ')} {issue.categoria}")
            linhas.append(f"       Coluna: {issue.coluna}")
            linhas.append(f"       {issue.descricao}")

            if issue.detalhes:
                for chave, valor in issue.detalhes.items():
                    if isinstance(valor, list) and len(valor) > 3:
                        linhas.append(f"       {chave}: {valor[:5]}... (mais {len(valor)-5})")
                    else:
                        linhas.append(f"       {chave}: {valor}")
            linhas.append("")

        return "\n".join(linhas)

    def sugerir_tratamentos(self) -> list:
        """
        Gera lista de etapas sugeridas (no formato do Pipeline) com base
        nos problemas encontrados.
        """
        etapas_sugeridas = []
        categorias = {i.categoria for i in self.issues}

        if "NULOS" in categorias:
            cols_nulos = [i.coluna for i in self.issues if i.categoria == "NULOS"]
            etapas_sugeridas.append({
                "operacao": "remover_colunas_vazias",
                "params": {"limiar": 1.0},
                "descricao": "Remover colunas totalmente vazias",
            })

        if "COLUNA_VAZIA" in categorias:
            etapas_sugeridas.append({
                "operacao": "remover_colunas_vazias",
                "params": {"limiar": 1.0},
                "descricao": "Remover colunas 100% vazias",
            })

        if "DUPLICATAS" in categorias:
            etapas_sugeridas.append({
                "operacao": "remover_duplicatas",
                "params": {"normalizar": True},
                "descricao": "Remover linhas duplicadas",
            })

        if "FORMATO_NUMERICO" in categorias:
            cols_formato = [i.coluna for i in self.issues if i.categoria == "FORMATO_NUMERICO"]
            etapas_sugeridas.append({
                "operacao": "numero_interno",
                "params": {"colunas": cols_formato, "casas": 2},
                "descricao": f"Converter textos numéricos para numero: {cols_formato}",
            })

        if "VALOR_SUSPEITO" in categorias:
            for issue in self.issues:
                if issue.categoria == "VALOR_SUSPEITO":
                    valores_suspeitos = list(issue.detalhes.get("valores", {}).keys())
                    mapeamento = {v: "" for v in valores_suspeitos}
                    etapas_sugeridas.append({
                        "operacao": "substituir_valores",
                        "params": {"colunas": [issue.coluna], "mapeamento": mapeamento},
                        "descricao": f"Limpar valores suspeitos em '{issue.coluna}': {valores_suspeitos}",
                    })

        if "TEXTO_ESPACO" in categorias:
            cols_espaco = [i.coluna for i in self.issues if i.categoria == "TEXTO_ESPACO"]
            etapas_sugeridas.append({
                "operacao": "normalizar_texto",
                "params": {"colunas": cols_espaco, "modo": "upper"},
                "descricao": f"Normalizar espacos e caixa em: {cols_espaco}",
            })

        if "FUZZY_MATCH" in categorias:
            for issue in self.issues:
                if issue.categoria == "FUZZY_MATCH":
                    pares = issue.detalhes.get("pares", [])
                    if pares:
                        mapeamento = {}
                        for a, b, _ in pares:
                            mapeamento.setdefault(a, []).append(b)
                        etapas_sugeridas.append({
                            "operacao": "unificar_valores",
                            "params": {"coluna": issue.coluna, "mapeamento": mapeamento},
                            "descricao": f"Unificar valores similares em '{issue.coluna}'",
                        })

        return etapas_sugeridas

    def para_arquivo(self) -> str:
        """Gera conteudo para salvar em arquivo .txt."""
        linhas = []
        linhas.append("=" * 70)
        linhas.append("  RELATORIO DE DIAGNOSTICO — ETL NGR-SEE")
        linhas.append("=" * 70)
        linhas.append("")
        linhas.append(self.resumo())
        linhas.append("")
        linhas.append("-" * 70)
        linhas.append("  DETALHAMENTO")
        linhas.append("-" * 70)
        linhas.append(self.detalhado())
        linhas.append("-" * 70)
        linhas.append("  SUGESTOES DE TRATAMENTO")
        linhas.append("-" * 70)
        etapas = self.sugerir_tratamentos()
        if etapas:
            for i, etapa in enumerate(etapas, 1):
                linhas.append(f"  {i}. {etapa['descricao']}")
        else:
            linhas.append("  Nenhuma acao sugerida.")
        linhas.append("=" * 70)
        return "\n".join(linhas)

File: src/diagnostics/report.ts
import { DiagnosticIssue, SuggestedStep } from "../types";

export class DiagnosticReport {
  private issues: DiagnosticIssue[];

  constructor(issues: DiagnosticIssue[]) {
    this.issues = issues;
  }

  public resumo(): string {
    if (!this.issues || this.issues.length === 0) {
      return "  Nenhum problema encontrado. Base parece saudavel.";
    }

    const criticos = this.issues.filter((i) => i.gravidade === "critico");
    const avisos = this.issues.filter((i) => i.gravidade === "aviso");
    const infos = this.issues.filter((i) => i.gravidade === "info");

    const linhas: string[] = [];
    linhas.push(`  Total de problemas: ${this.issues.length}`);
    linhas.push(`    Criticos: ${criticos.length}`);
    linhas.push(`    Avisos:   ${avisos.length}`);
    linhas.push(`    Info:     ${infos.length}`);

    const categorias: Record<string, number> = {};
    for (const i of this.issues) {
      categorias[i.categoria] = (categorias[i.categoria] || 0) + 1;
    }
    linhas.push("");
    linhas.push("  Por categoria:");
    for (const [cat, count] of Object.entries(categorias).sort((a, b) => b[1] - a[1])) {
      linhas.push(`    ${cat}: ${count}`);
    }

    return linhas.join("\n");
  }

  public detalhado(): string {
    if (!this.issues || this.issues.length === 0) {
      return "  Nenhum problema encontrado.";
    }

    const linhas: string[] = [];
    for (let i = 0; i < this.issues.length; i++) {
      const issue = this.issues[i];
      const icone = issue.gravidade === "critico" ? "[!!!]" : issue.gravidade === "aviso" ? "[ ! ]" : "[ i ]";
      linhas.push(`  ${String(i + 1).padStart(3, " ")}. ${icone} ${issue.categoria}`);
      linhas.push(`       Coluna: ${issue.coluna}`);
      linhas.push(`       ${issue.descricao}`);

      if (issue.detalhes) {
        for (const [chave, valor] of Object.entries(issue.detalhes)) {
          if (Array.isArray(valor) && valor.length > 3) {
            linhas.push(`       ${chave}: ${JSON.stringify(valor.slice(0, 5))}... (mais ${valor.length - 5})`);
          } else {
            linhas.push(`       ${chave}: ${typeof valor === "object" ? JSON.stringify(valor) : valor}`);
          }
        }
      }
      linhas.push("");
    }

    return linhas.join("\n");
  }

  public sugerirTratamentos(): SuggestedStep[] {
    const etapasSugeridas: SuggestedStep[] = [];
    const categorias = new Set(this.issues.map((i) => i.categoria));

    if (categorias.has("COLUNA_VAZIA") || categorias.has("NULOS")) {
      etapasSugeridas.push({
        operacao: "remover_colunas_vazias",
        params: { limiar: 1.0 },
        descricao: "Remover colunas 100% vazias",
      });
    }

    if (categorias.has("DUPLICATAS")) {
      etapasSugeridas.push({
        operacao: "remover_duplicatas",
        params: { normalizar: true },
        descricao: "Remover linhas duplicadas",
      });
    }

    if (categorias.has("NUMERO_PONTO_DECIMAL") || categorias.has("TIPO_TEXTO_NUMERO")) {
      const colsPonto = this.issues
        .filter((i) => i.categoria === "NUMERO_PONTO_DECIMAL" || i.categoria === "TIPO_TEXTO_NUMERO")
        .map((i) => i.coluna);
      if (colsPonto.length > 0) {
        const uniqueCols = Array.from(new Set(colsPonto));
        etapasSugeridas.push({
          operacao: "numero_interno",
          params: { colunas: uniqueCols, casas: 2 },
          descricao: `Converter textos numéricos para numero: ${uniqueCols.join(", ")}`,
        });
      }
    }

    if (categorias.has("TEXTO_ESPACO") || categorias.has("TEXTO_CASO")) {
      const colsEspaco = this.issues
        .filter((i) => i.categoria === "TEXTO_ESPACO" || i.categoria === "TEXTO_CASO")
        .map((i) => i.coluna);
      if (colsEspaco.length > 0) {
        const uniqueCols = Array.from(new Set(colsEspaco));
        etapasSugeridas.push({
          operacao: "normalizar_texto",
          params: { colunas: uniqueCols, modo: "upper" },
          descricao: `Normalizar espacos e caixa alta em: ${uniqueCols.join(", ")}`,
        });
      }
    }

    if (categorias.has("FUZZY_MATCH")) {
      for (const issue of this.issues) {
        if (issue.categoria === "FUZZY_MATCH") {
          etapasSugeridas.push({
            operacao: "unificar_valores",
            params: { coluna: issue.coluna, limiar: 0.82 },
            descricao: `Unificar valores similares em '${issue.coluna}'`,
          });
        }
      }
    }

    return etapasSugeridas;
  }

  public paraArquivo(): string {
    const linhas: string[] = [];
    linhas.push("=".repeat(70));
    linhas.push("  RELATORIO DE DIAGNOSTICO — ETL NGR-SEE");
    linhas.push("=".repeat(70));
    linhas.push("");
    linhas.push(this.resumo());
    linhas.push("");
    linhas.push("-".repeat(70));
    linhas.push("  DETALHAMENTO");
    linhas.push("-".repeat(70));
    linhas.push(this.detalhado());
    linhas.push("-".repeat(70));
    linhas.push("  SUGESTOES DE TRATAMENTO");
    linhas.push("-".repeat(70));
    const etapas = this.sugerirTratamentos();
    if (etapas.length > 0) {
      for (let i = 0; i < etapas.length; i++) {
        linhas.push(`  ${i + 1}. ${etapas[i].descricao}`);
      }
    } else {
      linhas.push("  Nenhuma acao sugerida.");
    }
    linhas.push("=".repeat(70));
    return linhas.join("\n");
  }
}
